File: frontend/src/types/timeline.ts
/**
 * TypeScript types for timeline (comments + activities)
 */

export interface User {
  id: string
  username: string
  email?: string
  avatar_url?: string | null
}

export interface TimelineComment {
  type: 'comment'
  id: string
  card_id: string
  author: User
  text: string
  created_at: string
  updated_at: string
}

export interface TimelineActivity {
  type: 'activity'
  id: string
  card_id: string
  user: User
  action: string
  description: string
  created_at: string
}

export type TimelineItem = TimelineComment | TimelineActivity

export interface TimelineResponse {
  items: TimelineItem[]
  total: number
  offset: number
  limit: number
}

export interface CommentCreateRequest {
  text: string
}

export interface CommentUpdateRequest {
  text: string
}

File: backend/app/schemas/timeline.py
"""Pydantic schemas for card timeline (combined comments and activities)."""

from datetime import datetime
from typing import Literal, Optional, Union
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.card_activity import ActivityAction


class TimelineUser(BaseModel):
    """Timeline entry user information."""

    id: UUID
    username: str
    avatar_url: Optional[str] = None

    class Config:
        """Pydantic config."""

        from_attributes = True


class TimelineComment(BaseModel):
    """Timeline comment entry."""

    type: Literal["comment"] = "comment"
    id: UUID
    card_id: UUID
    author: Optional[TimelineUser] = None
    text: str
    created_at: datetime
    updated_at: datetime

    class Config:
        """Pydantic config."""

        from_attributes = True


class TimelineActivity(BaseModel):
    """Timeline activity entry."""

    type: Literal["activity"] = "activity"
    id: UUID
    card_id: UUID
    user: Optional[TimelineUser] = None
    action: ActivityAction
    description: str
    created_at: datetime

    class Config:
        """Pydantic config."""

        from_attributes = True


TimelineItem = Union[TimelineComment, TimelineActivity]


class TimelineResponse(BaseModel):
    """Schema for paginated timeline."""

    items: list[TimelineItem] = Field(..., discriminator="type")
    total: int
    offset: int
    limit: int
